### ISSUE TITLE
CoreAPI should expose Host's stream methods

### ISSUE THREAD

**Issue by lthibault** [ipfs/go-ipfs#7145]
I am writing an application that uses go-libp2p as a library.  This application requires nodes to initiate and handle data streams to each other, in order to process RPC calls of various sorts.  Right now this achievable as follows:

```go
type ExtendedAPI struct {
    iface.CoreAPI
    stream StreamAPI
}

func newExtendedAPI(node *core.IpfsNode, api iface.CoreAPI) ExtendedAPI {
	return Foo{
		CoreAPI: api,
		ext:    node.PeerHost,
	}
}

func (api ExtendedAPI) Stream() StreamAPI { return api.ext }

func StreamAPI interface {
	SetStreamHandler(protocol.ID, network.StreamHandler)
	SetStreamHandlerMatch(protocol.ID, func(string) bool, network.StreamHandler)
	RemoveStreamHandler(protocol.ID)
	NewStream(context.Context,  peer.ID,  ...protocol.ID) (network.Stream, error)
}
```
Unless I'm overlooking some sort of footgun, this seems like the kind of thing that should be made available from `iface.CoreAPI`.  This would be a simple matter of adding the following method to the [`CoreAPI` struct](https://github.com/ipfs/go-ipfs/blob/v0.4.23/core/coreapi/coreapi.go#L49) :

```go
func (api *CoreAPI) Stream() coreiface.StreamAPI {
	return (*StreamAPI)(api)
}

// coreiface.StreamAPI declaration omitted, as it's trivial.
```

As usual, I'm happy to submit a PR if there's any interest.

**Comment by hsanjuan** [comment]
I'm not sure I follow. If using go-ipfs as a library you should have a libp2p `Host` at hand, and you can do stream handling with it. Why would we need to put it as part of the core API (particularly when stream handling is a libp2p thing, not an IPFS thing per se).

**Comment by lthibault** [comment]
>stream handling is a libp2p thing, not an IPFS thing per se

@hsanjuan I must have misread the intent of the `CoreAPI` interface.  My initial  assumption was that users should avoid interacting with the library _except_ via the core API (presumably for runtime-safety reasons).  It sounds like the intent is instead to provide the functionality from the IPFS command-line tool through a native Go interface.

tl;dr:  I can use `IpfsNode.PeerHost` directly.  This is fine.  Please disregard this issue!

(I'm guessing this also holds true for `IpfsNode.PubSub`?)

**Comment by hsanjuan** [comment]
>  It sounds like the intent is instead to provide the functionality from the IPFS command-line tool through a native Go interface.

In terms of stream handling, it is already part of the `Host` interface (https://godoc.org/github.com/libp2p/go-libp2p-core/host#Host). I'm not fully sure that parity between the command line and the Core API is a thing we want (given it would just be a wrapper over an existing interface). It seems this would just increase API surface unnecessarily.

Same for pubsub.  For the time being at least.
